### PR TITLE
Feat: add custom Router & useGoBack

### PR DESCRIPTION
**Pull request by HugoPerard**
~~Adding useSmartNavigate hook and SmartNavigateLink component, for manage smart navigation with automatic complete of state, thats permits to make smart goBack easily~~

~~⚠️ This feature need an update of "@testing-library/user-event" version from 13.5.0 to 14.2.0. It was not use in project until now, so that should not cause problem~~

- [x] Custom Router with state injection
- [x] Hook useGoBack
- [x] Storybook
- [x] Use useGoBack in the app

### CODE REVIEW

File: src/app/router/smartNavigate/utils.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { Location } from 'react-router-dom';
+
+/**
+ *
+ * @description build the complete current route from location
+ * @returns complete current route
+ */
+export const getCurrentRoute = (location: Location) =>
+  `${location.pathname}${location.search}`;
```
Comment (yoannfleurydev) [inline src/app/router/smartNavigate/utils.ts:9]: Missing `location.hash` ? https://developer.mozilla.org/en-US/docs/Web/API/Location

File: src/app/router/smartNavigate/useSmartNavigate.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export type SmartNavigateState = {
+  from: string;
+};
+
+/**
+ * @description Hook for manage goBack with an automatic state added when navigate
+ *
+ * @returns smartNavigate: navigate override that add *from* value to state
+ * @returns smartGoBack: goBack using state *from* value, or fallbackRoute param if there is no *from*
```
Comment (yoannfleurydev) [inline src/app/router/smartNavigate/useSmartNavigate.ts:18]: Are you sure about this documentation as you return only 1 object.

File: src/app/router/smartNavigate/SmartNavigateLink.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { FC } from 'react';
+
+import { Link, LinkProps, useLocation } from 'react-router-dom';
+
+import { getCurrentRoute } from '@/app/router/smartNavigate/utils';
+
+/**
+ *
+ * @description react-router-dom Link overcoat that add state *from* value for manage smart navigation
```
Comment (yoannfleurydev) [inline src/app/router/smartNavigate/SmartNavigateLink.tsx:9]:
```suggestion
 * @description react-router-dom Link state override that add `from` value to handle smart navigation.
```

**Review by yoannfleurydev — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -73,7 +73,7 @@
     "@testing-library/jest-dom": "5.16.3",
     "@testing-library/react": "12.1.4",
     "@testing-library/react-hooks": "7.0.2",
-    "@testing-library/user-event": "13.5.0",
+    "@testing-library/user-event": "14.2.0",
```
Comment (Jessy-BAER) [inline package.json:76]: Are you sure your need it for your feature ?

File: src/app/router/useGoBack.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { useLocation, useNavigate } from 'react-router-dom';
+
+export const useGoBack = (defaultFallback?: string) => {
```
Comment (Jessy-BAER) [inline src/app/router/useGoBack.ts:3]: You must add comments on this custom hook

File: src/app/router/Router.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  createBrowserHistory,
+  createMemoryHistory,
+} from 'history';
+import { Router as ReactRouter, RouterProps } from 'react-router-dom';
+
+const removeBasenameFromPathname = (basename: string, pathname: string) => {
+  return pathname.replace(new RegExp(`^${basename}`, 'g'), '');
+};
+
+export const Router = ({
```
Comment (Jessy-BAER) [inline src/app/router/Router.tsx:15]: It's a new important component, please add comments on it

**Comment by HugoPerard** [comment]
Sonar detects 1 code smell for an unnecessary assertion because that doesn't change type, but if I don't do it there is a typescript error. Because default type is unknown, not any

**Comment by yoannfleurydev** [comment]
Related to `react-router-dom` that we do not use anymore, I'm closing this PR.

**Pull request closed without merging**